hail/src/main/scala/is/hail/expr/ir/lowering/LowerBlockMatrixIR.scala

```diff
@@ -965,6 +965,24 @@ object LowerBlockMatrixIR {
         val Array(keepRow, keepCol) = keep
         lower(child).filter(keepRow, keepCol, x.typ, ib)
 
+      case x@BlockMatrixDensify(child) =>
+        val Some((rowPos, rowIdx)) = x.typ.sparsity.definedBlocksCSCIR(x.typ.nColBlocks)
+        lower(child).withSparsity(ib.memoize(rowPos), ib.memoize(rowIdx), ib, x.typ)
+        /*
+      case x@BlockMatrixSparsify(child, sparsifier) =>
+        val Some((rowPos, rowIdx)) = x.typ.sparsity.definedBlocksCSCIR(x.typ.nColBlocks)
+        val loweredChild = lower(child).withSparsity(ib.memoize(rowPos), ib.memoize(rowIdx), ib, x.typ, isSubset = true)
+        sparsifier match {
+          // these cases are all handled at the type level
+          case BandSparsifier(blocksOnly, _, _) if (blocksOnly) => loweredChild
+          case RowIntervalSparsifier(blocksOnly, _, _) if (blocksOnly) => loweredChild
+          case PerBlockSparsifier(_) => loweredChild
+
+          case BandSparsifier(_, l, u) => ???
+          case RowIntervalSparsifier(_, starts, stops) => ???
+          case RectangleSparsifier(rectangles) => ???
```

Comment:
@patrick-schultz I think I need some advice.

It's feeling like for these cases, I'll need to re-implement the zero methods of BlockMatrix like https://github.com/chrisvittal/hail/blob/2b358c8f355f19e3aafba86ec03eab616bd0d6cb/hail/src/main/scala/is/hail/linalg/BlockMatrix.scala#L492-L544

I think I know what I need to do to go about adding data to the context so that I can use it for filtering in the stage. However, how do I go about zeroing a range for an NDArray?

Reply:
I think we need new functionality for that. I'd probably just make them registered functions in `NDArrayFunctions`.